app/gui2/src/components/GraphEditor/widgets/WidgetFileBrowser.vue

```diff
@@ -0,0 +1,119 @@
+} from '@/components/GraphEditor/widgets/WidgetSelection.vue'
+import { Score, WidgetInput, defineWidget, widgetProps } from '@/providers/widgetRegistry'
+import { useGraphStore } from '@/stores/graph'
+import type { RequiredImport } from '@/stores/graph/imports'
+import { Ast } from '@/util/ast'
+import { ArgumentInfoKey } from '@/util/callTree'
+import { computed } from 'vue'
+
+const props = defineProps(widgetProps(widgetDefinition))
+const graph = useGraphStore()
+
+const insertAsFileConstructor = computed(() => {
+  const reprType = props.input[ArgumentInfoKey]?.info?.reprType
+  if (reprType == null) return false
+  const textTypePos = reprType.indexOf(TEXT_TYPE)
+  const fileTypePos = reprType.indexOf(FILE_TYPE)
+  // If both types are present, check if TEXT_TYPE comes after FILE_TYPE.
+  if (fileTypePos !== -1 && (textTypePos === -1 || textTypePos > fileTypePos)) {
+    return true
+  } else {
+    return false
+  }
+})
+const strictlyFile = computed(() => props.input.dynamicConfig?.kind === 'File_Browse')
+const strictlyDirectory = computed(() => props.input.dynamicConfig?.kind === 'Folder_Browse')
+const label = computed(() => (strictlyDirectory.value ? 'Choose directory…' : 'Choose file…'))
+
+const FILE_CONSTRUCTOR = FILE_TYPE + '.new'
+const FILE_SHORT_CONSTRUCTOR = 'File.new'
+
+function makeValue(edit: Ast.MutableModule, useFileConstructor: boolean, path: string): Ast.Owned {
+  if (useFileConstructor) {
+    const arg = Ast.TextLiteral.new(path, edit)
+    const requiredImport = {
+      kind: 'Unqualified',
+      from: FILE_MODULE,
+      import: 'File',
+    } as RequiredImport
+    const conflicts = graph.addMissingImports(edit, [requiredImport])
+    const constructor = conflicts != null ? FILE_CONSTRUCTOR : FILE_SHORT_CONSTRUCTOR
+    const constructorAst = Ast.PropertyAccess.tryParse(constructor, edit)
+    if (constructorAst == null) {
+      throw new Error(`Failed to parse constructor as AST: ${constructor}`)
+    }
+    return Ast.App.new(edit, constructorAst, undefined, arg)
+  } else {
+    return Ast.TextLiteral.new(path, edit)
+  }
+}
+
+const item = computed<CustomDropdownItem>(() => ({
+  label: label.value,
+  onClick: async () => {
+    const kind =
+      strictlyDirectory.value ? 'directory'
+      : strictlyFile.value ? 'file'
+      : 'any'
+    const selected = await window.fileBrowserApi.openFileBrowser(kind)
+    if (selected != null && selected[0] != null) {
+      const edit = graph.startEdit()
+      const value = makeValue(edit, insertAsFileConstructor.value, selected[0])
+      props.onUpdate({
+        edit,
+        portUpdate: {
+          value,
+          origin: props.input.portId,
+        },
+      })
+    }
+  },
+}))
+
+const innerWidgetInput = computed(() => {
+  const input = props.input
+  if (input[CustomDropdownItemsKey] != null) {
+    input[CustomDropdownItemsKey].push(item.value)
+  } else {
+    input[CustomDropdownItemsKey] = [item.value]
+  }
+  return input
+})
```

Comment:
This computed mutates the input array if its already present. That can lead to unexpected behavior, since computed values are meant to not have side effects (for the most part).

```suggestion
const innerWidgetInput = computed(() => {
  const existingItems = props.input[CustomDropdownItemsKey] ?? []
  return {
    ...props.input,
    [CustomDropdownItemsKey]: [...existingItems, item.value]
  }
})
```

Reply:
Thanks for catching. I’ve already tried to avoid this, but forgot that `const input = props.input` doesn’t make a copy! Although it removes a linter warning, which I guess confused me.